src/main/java/org/folio/rest/impl/RequestsAPI.java

```diff
@@ -134,9 +136,28 @@ private Errors samePositionInQueueError(Request request) {
   }
 
   private boolean isSamePositionInQueueError(AsyncResult<Response> reply) {
-    return reply.succeeded()
-      && reply.result().getStatus() == 400
-      && RequestsApiUtil
-      .hasSamePositionConstraintViolated(reply.result().getEntity().toString());
+    String message = null;
+
+    if (reply.succeeded() && reply.result().getStatus() >= 400 &&
+      reply.result().getStatus() < 600 && reply.result().hasEntity()) {
```

Comment:
Why is this a check between 400 and 600, are client and server errors treated the same by PgUtil?

Reply:
This is just an attempt to make it less fragile. That said, I am fine with making it very specific to exact error codes.

The problem seems to be that sometimes we get 400, 422, and 500 of which, apparently, was all returning the same codes prior to the update based for the original tests.

There is also a situation where the deprecated MyPgUtil upsert is used and that operates with different return objects, codes, and messages than PgUtil.

Reply:
> This is just an attempt to make it less fragile. That said, I am fine with making it very specific to exact error codes.
> 
> The problem seems to be that sometimes we get 400, 422, and 500 of which, apparently, was all returning the same codes prior to the update based for the original tests.

I don't think we need to be too specific. I think we do want to separate 4xx and 5xx errors.

Does the same underlying cause (e.g. multiple open loans or request in the same position) sometimes cause PgUtil to report a 4xx and 5xx error?

> There is also a situation where the deprecated MyPgUtil upsert is used and that operates with different return objects, codes, and messages than PgUtil.

Can the `MyPgUtil` use be replaced with `PgUtil` within this pull request?

Reply:
> Can the MyPgUtil use be replaced with PgUtil within this pull request?

It looks to me like doing that would incur some significant changes, possibly breaking changes. The MyPgUtil is being used only for the Upserts, which as far as I can tell were not ported over to PgUtil.

I can think of a few approaches here:
1. Implement upserts in PgUtil, but I imagine they didn't implement it for some reason.
1. Implement a select, check its response, and then conditionally call insert/update as appropriate.
1. Redesign something to avoid the need for an upsert, if this is even possible.
1. It is already marked as deprecated, some parts of these tests could be somehow marked as requiring changes relating to the deprecating functions (commenting out affected tests or implementing temporary solutions, for example).

Reply:
> > Can the MyPgUtil use be replaced with PgUtil within this pull request?
> 
> It looks to me like doing that would incur some significant changes, possibly breaking changes. The MyPgUtil is being used only for the Upserts, which as far as I can tell were not ported over to PgUtil.
> 
> I can think of a few approaches here:
> 
> 1. Implement upserts in PgUtil, but I imagine they didn't implement it for some reason.

Please can you ask the person who deprecated the upsert method in MyPgUtil what the recommended alternative approach is

> 2. Implement a select, check its response, and then conditionally call insert/update as appropriate.

That is what has been done in other places.

> 3. Redesign something to avoid the need for an upsert, if this is even possible.

I don't know enough about the context to answer that.

> 4. It is already marked as deprecated, some parts of these tests could be somehow marked as requiring changes relating to the deprecating functions (commenting out affected tests or implementing temporary solutions, for example).

Is MyPgUtil part of the production or test code?

If it is part of the production code, then are you proposing changing the tests to reflect changes to error message?

Reply:
> Is MyPgUtil part of the production or test code?

MyPgUtil is part of production (`org.folio.rest.persist.MyPgUtil`).

In case 4, I am proposing **temporarily** changing the test code to reflect changes.
This would, in theory, allow for a new ticket to be opened up so that the resolution of the deprecated Upsert (and related necessary changes discovered in this PR) can be performed in completely separate ticket, branch, and PR.

Reply:
> > Is MyPgUtil part of the production or test code?
> 
> MyPgUtil is part of production (`org.folio.rest.persist.MyPgUtil`).
> 
> In case 4, I am proposing **temporarily** changing the test code to reflect changes.
> This would, in theory, allow for a new ticket to be opened up so that the resolution of the deprecated Upsert (and related necessary changes discovered in this PR) can be performed in completely separate ticket, branch, and PR.

Agreed, let's create new JIRA issue for the technical debt and change the test code where the behaviour remains intact yet the message has changed.

Reply:
The collection rtype assumes that a PUT is an update only, not an upsert, and should return 404 when the record doesn't exists: https://github.com/folio-org/raml/blob/raml1.0/rtypes/item-collection.raml#L65
https://s3.amazonaws.com/foliodocs/api/mod-circulation-storage/request-storage.html#request_storage_requests__requestid__put
I propose to change the upsert to an update to match the API. This is a bug fix and not a breaking API change. Clients can (and usually do) use the POST API for inserts.
This is the reason why there is no need to have an upsert method in PgUtil.

In addition the current upsert implementation returns 204 on insert, it should return 201. This wrong status code is the reason why I marked MyPgUtil as deprecated.

Reply:
I think, where possible, we should preserve the current behaviour of mod-circulation-storage prior to this upgrade (excluding specific error messages)